Implement the `join_inputs_design` into the current SDA workflows.

DongchenZ opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [x] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
Bug fixes for the missing ensemble.size error.

mdietze reviewed (commented):
FYI, this provides an immediate fix, but really you should be passing in the new input_design object too, not just the new ensemble.size variable. That change is needed to get the desired result of sampling jointly once at the start of the SDA and then keeping that design as the SDA iterates.

mdietze reviewed (changes requested):
Just checked are was reminded that input_design is now the first argument to write.ensemble.configs and it's not optional, so adding that fix is actually required in this PR, not just a good next step for a future PR. 

https://github.com/PecanProject/pecan/blob/ccc5135da7a0007560821cac7640d3d59c89afde/modules/uncertainty/R/ensemble.R#L228

I'd recommend that you add a call to modules/uncertainty/R/generate_joint_ensemble_design once in your code early on, that you save the design to restart, and that you read that design during restart rather than generating a new one (i.e., that if you have an "if restart" option, then the input_design is generated in the 
"else")

Commit:
Implement joint input sampling in the SDA workflows.

Commit:
Update the joint inputs sampling function.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Update the ensemble configs function.

Commit:
Clean the multisite sda function.

Commit:
Revert the changes.

Commit:
Revert changes.

Commit:
revert changes.

Commit:
Revert changes.

Commit:
Add documentation and bug fixes.

Commit:
Update document.

Commit:
Update the function to work with the newest functions.

Commit:
Remove the samples argument from the function.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -44,19 +41,22 @@ sda.enkf.multisite <- function(settings,
-                                            TimeseriesPlot = FALSE,
-                                            debug = FALSE,
-                                            pause = FALSE,
-                                            Profiling = FALSE,
```
Removed unnecessary flags.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -44,19 +41,22 @@ sda.enkf.multisite <- function(settings,
+                               control=list(TimeseriesPlot = FALSE,
                                             OutlierDetection=FALSE,
-                                            parallel_qsub = TRUE,
+                                            parallel_qsub = FALSE,
```
Set the `parallel_qsub` as `FALSE` because we can't use both local and remote `qsub` for the model execution.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -44,19 +41,22 @@ sda.enkf.multisite <- function(settings,
                                             run_parallel = TRUE,
-                                            MCMC.args = NULL),
+                                            MCMC.args = NULL,
+                                            local.execution = TRUE),
```
Add the `local.execution` when we just want to run the jobs locally.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -44,19 +41,22 @@ sda.enkf.multisite <- function(settings,
+    PEcAn.logger::logger.info("Please make sure you have only one of the following control options turning on.")
+    PEcAn.logger::logger.info("parallel_qsub or local.execution.")
+    return(0)
+  }
```
Make sure we only set one of the two flags as `TRUE` as mentioned earlier.

mdietze replied:
rather than having both variables, would it be better to define one variable able to take on more than two options e.g. `execution = c("local", "qsub", "parallel_qsub")`

DongchenZ replied:
Fixed.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -65,15 +65,14 @@ sda.enkf.multisite <- function(settings,
   }else{
     restart_flag = FALSE
   }
+  # register parallel nodes.
```
Add comment.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -84,50 +83,23 @@ sda.enkf.multisite <- function(settings,
-  censored.data<-ifelse(is.null(settings$state.data.assimilation$censored.data), 
-                        TRUE, 
-                        settings$state.data.assimilation$censored.data %>% 
-                          as.logical)
```
Remove the traditional settings, as we have already set up `MCMC.args` in the arguments.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -84,50 +83,23 @@ sda.enkf.multisite <- function(settings,
-    processvar <- FALSE
-  }
-  Localization.FUN <- settings$state.data.assimilation$Localization.FUN # localization function
-  scalef <- settings$state.data.assimilation$scalef %>% as.numeric() # scale factor for localization
```
Remove unnecessary variable definitions.

mdietze replied:
How are these variables now unnecessary? How do you control localization if these are deleted?

DongchenZ replied:
1. Those two variables are not used anywhere in the script.
2. The localization will be applied within the `analysis` function instead of being here. eg., `analysis_block` function:https://github.com/PecanProject/pecan/blob/6d7a913dd9c5f6f3f992cbe3f9e3f263cd56bb6f/modules/assim.sequential/R/Analysis_sda_block.R#L144;
and the traditional `analysis` function: https://github.com/PecanProject/pecan/blob/6d7a913dd9c5f6f3f992cbe3f9e3f263cd56bb6f/modules/assim.sequential/R/Analysis_sda_multiSite.R#L150

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -84,50 +83,23 @@ sda.enkf.multisite <- function(settings,
-  
-  ##### Creating matrices that describe the bounds of the state variables
-  ##### interval is remade everytime depending on the data at time t
-  ##### state.interval stays constant and converts new.analysis to be within the correct bounds
```
Simplify comments.

mdietze replied:
I'd revert, no need to remove useful information

DongchenZ replied:
fixed.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -136,7 +108,7 @@ sda.enkf.multisite <- function(settings,
   if(multi.site.flag){
     conf.settings<-settings
-    site.ids <- conf.settings$run %>% purrr::map('site') %>% purrr::map('id') %>% base::unlist() %>% base::as.character()
+    site.ids <- conf.settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('id') %>% base::unlist() %>% base::as.character()
```
I should revert this change.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -218,7 +184,7 @@ sda.enkf.multisite <- function(settings,
     dir.create(paste0(settings$outdir, "/Extracted_met/"))
   }
-  
+  PEcAn.logger::logger.info("Splitting mets!")
```
Add progress report messaging.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -295,8 +260,15 @@ sda.enkf.multisite <- function(settings,
+          if (!exists("input_design")) {
+            input_design <- NULL
+          }
+          # wrtting configs for each settings - this does not make a difference with the old code
```
Will remove this if-control and move the one that handles error messaging and returns up here (or earlier).

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -343,65 +315,31 @@ sda.enkf.multisite <- function(settings,
-        input_path <- conf.settings[[s]]$run$inputs[[tolower(input_tag)]]$path
-        inputs[[s]][[input_tag]]$ids<-inputs[[random_site]][[input_tag]]$ids
-        inputs[[s]][[input_tag]]$samples<- input_path[inputs[[random_site]][[input_tag]]$ids]
-      }
```
Remove this whole section because we don't need the inputs/samples, which will be created during the `write.configs` function later.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -343,65 +315,31 @@ sda.enkf.multisite <- function(settings,
+    input_design <- generate_joint_ensemble_design(settings = settings[[1]], 
+                                                   ensemble_samples = ensemble.samples, 
+                                                   ensemble_size = nens)[[1]]
   }
```
Add the function call to create the `input_design`.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+                                                      } %>% stats::setNames(site.ids)
+        parallel::stopCluster(cl)
+        foreach::registerDoSEQ()
+        # update the file paths of different inputs when t = 1.
```
Here, we switch from `furrr` to the `foreach`. This is because, for some reason, the `furrr` doesn't return the correct samples returned from the `write.config` function. For example, only the `met` and `parameter` will be returned if we sample all the `met`, `poolinitcond`, and `parameter` inputs.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+        # update the file paths of different inputs when t = 1.
+        if (t == 1) {
+          inputs <- out.configs %>% purrr::map(~.x$samples)
+        }
```
Here, we need to grab the correct inputs from the `write.config` function when `t=1`, and save them for later use (when `t>1`, we will be relying on them).

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+        runs.tmp <- file.path(rundir, ensemble.ids)
+        # start model runs.
+        PEcAn.logger::logger.info("Running models!")
+        # if we want to submit jobs through the combined job file.
```
Update the way we create the `run.id` file.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+            system(cmd, intern = F, ignore.stdout = T, ignore.stderr = T)
+          }, .progress = F)
+        } else {
+          # if we want to submit jobs through the regular job submission function.
```
Add the local execution part.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+                         t = t, 
+                         var.names = var.names, 
+                         my.read_restart = my.read_restart,
+                         restart_flag = restart_flag)
```
Remove the error detections as they are unnecessary now.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -440,100 +377,110 @@ sda.enkf.multisite <- function(settings,
+          colnames(temp) <- paste0(var.names, ".", i)
+          return(temp)
+        }) %>% 
+          dplyr::bind_cols() %>%
```
Convert the forecasts to a data frame.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -543,33 +490,27 @@ sda.enkf.multisite <- function(settings,
+                            nburnin = 5e4)
+        } else {
+          MCMC.args <- control$MCMC.args
         }
```
Grab the `MCMC.args` from the arguments or set it up.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -677,29 +617,6 @@ sda.enkf.multisite <- function(settings,
-        }
-        if (control$debug) browser()
-        if (control$pause) readline(prompt="Press [enter] to continue \n")
       } else {
```
Remove the trace part as they no longer be informative.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -738,7 +655,12 @@ sda.enkf.multisite <- function(settings,
+          analysis <- as.data.frame(mvtnorm::rmvnorm(as.numeric(nrow(X)), mu.a, Pa, method = "svd"))
+        } else {
+          analysis <- enkf.params[[obs.t]]$analysis
+        }
```
The new analysis function provides samples of analysis ensembles, which are calculated across blocks, and it is faster than sampling everything together, so we don't have to redo the job.

DongchenZ commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -226,7 +225,7 @@ get.ensemble.samples <- function( ensemble.size, pft.samples, env.samples,
 ##' @author David LeBauer, Carl Davidson, Hamze Dokoohaki
 write.ensemble.configs <- function(input_design , ensemble.size, defaults, ensemble.samples, settings, model, 
-                                   clean = FALSE, write.to.db = TRUE, restart = NULL, samples = NULL, rename = FALSE) {
+                                   clean = FALSE, write.to.db = TRUE, restart = NULL, rename = FALSE) {
```
Remove the `samples` argument as it is no longer needed.

DongchenZ commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -320,22 +319,18 @@ for (input_tag in names(settings$run$inputs)) {
+        samples[[input_tag]] <- list(
+          samples = lapply(input_indices, function(idx) input_paths[[idx]])
+        )
+      }
```
Remove the `samples` argument as it is no longer needed.

DongchenZ commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -515,6 +510,8 @@ for (input_tag in names(settings$run$inputs)) {
       input_list <- list()
       for (input_tag in names(inputs)) {
+        # if it's the parameter list, skip.
+        if (input_tag == "parameters") next
```
Make sure we are not sending parameters for the paths, as the new `inputs_design` does include the `parameter` field.

DongchenZ commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -4,19 +4,20 @@
-#' @param sobol for activating sobol
 #' @param ensemble_size Integer specifying the number of ensemble members
+#' @param ensemble_samples list of ensemble parameters across PFTs. The default is NULL.
+#' @param sobol for activating sobol
```
Add the `ensemble_samples` argument to ensure we can also pass the parameters directly.

mdietze replied:
I'm not following the logic of why we're making ensemble_samples an internal requirement for the input_design. I don't see where/how we're actually using this anywhere. The sampling of parameters seems to be based solely on knowing the ensemble_size

DongchenZ commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -51,29 +52,34 @@ generate_joint_ensemble_design <- function(settings,
+      ens.sample.method
+    )
+    samples.file <- file.path(settings$outdir, "samples.Rdata")
+  }
```
Add `if-control` for the case where we don't assign the `ensemble_samples` argument.

mdietze replied:
Could you clarify the argument for when it makes sense to assume that we **have** ensemble_samples already? In other words, in what cases would you want/need to recycle a previous ensemble_samples rather than generate a new sample from the posteriors?

DongchenZ replied:
done

DongchenZ commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -51,29 +52,34 @@ generate_joint_ensemble_design <- function(settings,
   } else {
-    PEcAn.logger::logger.error(samples.file, "not found, this file is required")
+    PEcAn.logger::logger.warn("ensemble.samples not found in samples.Rdata")
   }
```
Add `if-control` for the case where we don't assign the `ensemble_samples` argument.

DongchenZ reviewed (commented):
Here, I have added a couple of bug fixes related to the new `joint_inputs_design` feature to the current SDA workflows. In addition, I also made a couple of modifications to the `sda_multisite` script to remove unnecessary/redundant parts, better aligning it with the current workflows. The revised SDA workflow was tested and found to work on the NEON-scale experiments.

Commit:
Add changes to the sda workflow.

Commit:
Add namespace.

Commit:
Bug fix.

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -289,14 +254,26 @@ sda.enkf.multisite <- function(settings,
       #TODO: there should be an easier way to do this than to rerun write.ensemble.configs
       restart.list <- vector("list", length(conf.settings))
+      # make sure we have the input_design variable before running the write configuration function.
+      if (!exists("input_design")) {
```
I'm confused by this bit. First, if input_design isn't an argument, and this is the non-restart part of the if/else, then input_design will never exist at this point. Second, if input_design doesn't exist, then why not just generate it instead of throwing an error?

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -343,65 +320,31 @@ sda.enkf.multisite <- function(settings,
-      }
+  #create params object using samples generated from TRAITS functions
+  if(restart_flag){
+    new.params <- new.params
```
This line doesn't do anything

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -343,65 +320,31 @@ sda.enkf.multisite <- function(settings,
+  if(restart_flag){
+    new.params <- new.params
+  } else {
+    if(!file.exists(file.path(settings$outdir, "samples.Rdata"))) PEcAn.logger::logger.severe("samples.Rdata cannot be found. Make sure you generate samples by running the get.parameter.samples function before running SDA.")
```
`get.parameter.samples` is called WITHIN `generate_joint_input_design`. This shouldn't need to be generated ahead of time. What the input design function needs is to be able to find the posterior files from the settings object.

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -543,33 +493,27 @@ sda.enkf.multisite <- function(settings,
+        }
+        #initialize MCMC arguments.
+        if (is.null(control$MCMC.args)) {
+          MCMC.args <- list(niter = 1e5,
```
drop this here and just make these the defaults directly in the function args

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -756,7 +681,7 @@ sda.enkf.multisite <- function(settings,
       ###-------------------------------------------------------------------###---- 
       Viz.output <- list(settings, obs.mean, obs.cov) #keeping obs data and settings for later visualization in Dashboard
-      
+      # save SDA outputs.
```
input_design needs to be added to the save!

mdietze commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -51,29 +52,34 @@ generate_joint_ensemble_design <- function(settings,
-      # You may want to flatten or select indices per trait
-      design_list[["param"]] <- seq_len(ensemble_size)
+  # if we don't have the parameters from the outside. 
+  if (is.null(ensemble_samples)) {
```
Why have to consecutive if(is.null(ensemble_samples) blocks? why can't these be combined into one block?

mdietze commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -51,29 +52,34 @@ generate_joint_ensemble_design <- function(settings,
+    design_list[["param"]] <- seq_len(ensemble_size)
   } else {
-    PEcAn.logger::logger.error(samples.file, "not found, this file is required")
+    PEcAn.logger::logger.warn("ensemble.samples not found in samples.Rdata")
```
why drop this from error to warn?

Commit:
Apply Mike's comments.

Commit:
Add the merge nc feature.

Commit:
Improve scripts.

Commit:
Bug fix.

Commit:
Update the parallel registration.

Commit:
Fix the dimension error when YN equals one.

Commit:
Update the parallel registration. And add the else control when it's a free run.

Commit:
Convert from SoilMoist to SoilMoistFrac.

Commit:
Add the recursive file deletion.

Commit:
Revert changes to resolve conflicts.

Commit:
revert change.

Commit:
Pull from pecan develop.

Merge branch 'develop' of https://github.com/PecanProject/pecan into SDA_data

# Conflicts:
#	modules/assim.sequential/R/sda.enkf_MultiSite.R

Commit:
change by make.

Commit:
Combine previous changes with the debias implementation.

Commit:
Merge the debiasing feature into the parallel SDA workflow.

Commit:
Update document.

Commit:
Update document.

Commit:
Add the debiasing feature to the parallel job submission.

Commit:
map forecast results to the variable boundary.

Commit:
Update comments.

Commit:
update observation paths.

Commit:
Update computation configuration.

Commit:
Bug fix.

Commit:
Bug fix.

Commit:
Revert change.

Commit:
Revert change.

Commit:
Add debias code.

Commit:
Move the previous debias script to the inst folder.

Commit:
remove previous debias functions.

Commit:
Add the main SDA debias workflow.

Commit:
Add the new debias workflow to the multi-site version.

Commit:
Add the new bias correction workflow to the sda_local script.

Commit:
Add the new SDA debiasing workflow to the SDA job submission function.

Commit:
Move out of the if control.

Commit:
Bug fix.

Commit:
Bug fix

Commit:
Remove the condition for the wrapping where we have more than 20 characters.

DongchenZ commented on base/logger/R/logger.R:
```diff
@@ -146,7 +146,7 @@ logger.message <- function(level, msg, ..., wrap = TRUE) {
       "latin-ascii"
     )
-    if (nchar(long.msg) > 20 && wrap) {
+    if (wrap) {
```
The hard-coded length in the if control should be removed. The `wrap` argument should fully define it.

DongchenZ commented on models/stics/NAMESPACE:
```diff
@@ -2,5 +2,6 @@
 
 export(met2model.STICS)
 export(model2netcdf.STICS)
+export(pecan2stics)
```
This is generated by building the existing pecan files.

DongchenZ commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -21,11 +21,12 @@ analysis_sda_block <- function (settings, block.list.all, X, obs.mean, obs.cov,
   }
+  cores <- cores - 1
+  # if we only have one CPU.
+  if (cores < 1) cores <- 1
```
Update the parallel registration.

DongchenZ commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -17,59 +17,47 @@
+#' `qsub_parallel`, where we first combine jobs and submit them into the SCC.
+#' @param debias List: R list containing the covariance directory and the start year.
+#' covariance directory should include GeoTIFF files named by year.
+#' start year is numeric input which decide when to start the debiasing feature.
```
Removed the previous `debias` workflow and replaced it with the new workflow.

Commit:
Update the runner script.

DongchenZ reviewed (commented):
The newest commits provide the new `debias` workflow (which has already been implemented in the current SDA workflows), and the old script has been moved to the inst folder. This is cleaner and easier to understand, which could also benefit any future development of the debiasing algorithm.

Commit:
remove arg.

Commit:
Update script.

Commit:
Revert "Move out of the if control."

This reverts commit 77bda481f7bc9a4065d8b05a16145ff03f4b2201.

Commit:
Resolve conflicts.

Commit:
Revert back.

Commit:
Revert to using the new debias workflow.

Commit:
Update document.

Commit:
Update document.

Commit:
Update the runner script.

Commit:
Update the bias correction function.

Commit:
Update the usage of the bias correction workflow.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Update observation paths.

Commit:
Update namespace.

Commit:
Merge branch 'SDA_data' of https://github.com/DongchenZ/pecan into SDA_data

Commit:
Modify the test to match the change in the logger.message function.

Commit:
Change the default value of wrap in the logger message function.

Commit:
Revert back.

DongchenZ commented on modules/data.remote/R/SMAP_SMP_prep.R:
```diff
@@ -88,15 +88,15 @@ SMAP_SMP_prep <- function(site_info, start_date, end_date, time_points,
   }
   PEcAn.logger::logger.info("SMAP SMP Prep Completed!")
-  list(SMP_Output = SMAP_Output, time_points = time_points, var = "SoilMoist")
+  list(SMP_Output = SMAP_Output, time_points = time_points, var = "SoilMoistFrac")
```
Correct the variable name from `SoilMoist` to ` SoilMoistFrac`.

Commit:
Update runner script.

Commit:
Modify configuration.

Commit:
Add the prefix argument to customize the desired output folder.

Commit:
Update configurations.

Commit:
Update the SDA runner script.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Merge branch 'SDA_data' of https://github.com/DongchenZ/pecan into SDA_data

infotroph commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -85,7 +96,5 @@ generate_joint_ensemble_design <- function(settings,
 
-
-
   return(list(X = design_matrix))
```
Is it important that this be a list? All the uses I've seen so far immediately subset the result to only store the X component, so would it make sense to return `design_matrix` unwrapped in the first place?

mdietze replied:
It is important so that the sobol version works correctly. This ensures that regardless of whether the sobol or non-sobol version is called that the output is a list that includes the design as X. In the sobol version the list includes additional info beyond just X that's required by the function that does the sobol index calculations, but not required to do the runs themselves

infotroph replied:
👍 Can you add this to the function docs, please? I think it'd be enough to list in the @return block the possible items for the Sobol case and mention that in the non-sobol case the list only contains X

DongchenZ replied:
Fixed.

infotroph replied:
I was picturing something that appeared in the `@return` section of the Roxygen documentation. Maybe something like...

```
#' @return a list with at least component `X` (a dataframe containing the computed design matrix).
#'   If `sobol = TRUE`, the list will be a `sensitivity::soboljansen()` result and will contain the components documented therein.
```

DongchenZ replied:
Done

infotroph commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -17,59 +17,47 @@
 #' @param pre_enkf_params Used for passing pre-existing time-series of process error into the current SDA runs to ignore the impact by the differences between process errors.
-#' @param ensemble.samples Pass ensemble.samples from outside to avoid GitHub check issues.
+#' @param ensemble.samples list of ensemble parameters across PFTs. Default is NULL.
 #' @param control   List of flags controlling the behavior of the SDA. 
```
This would be easier to read as a bulleted list -- consider making this line be something like ` @param control   List of flags controlling the behavior of the SDA. See details`, then moving all the flag descriptions on lines 22-34 into their own section of the details?

Commit:
Bug fixes for the indexing error.

Commit:
Fix the issue when X is a data frame.

Commit:
Add the adjustment to the analysis block function.

Commit:
Add the adjustment to the function.

Commit:
Add the residual lag to the SDA ML debiasing workflow.

Commit:
Use pre-defined soil depth for the soil physics calculation.

Commit:
Resolve conflicts.

Merge branch 'develop' of https://github.com/PecanProject/pecan into SDA_data

# Conflicts:
#	modules/uncertainty/R/ensemble.R

Commit:
Add comments for the reason of returning the list object.

Commit:
Fix the logic where we already generated analysis from the analysis block function.

Commit:
Remove the ensemble.samples argument.

Commit:
Update documentation.

mdietze commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -57,35 +56,10 @@ generate_joint_ensemble_design <- function(settings,
   }
-  
-  # Sample parameters if we don't have it.
-  if (is.null(ensemble_samples)) {
```
It needs to be detected a different way, but this function is still the one responsible for running get.parameter.samples if there's not already a samples.Rdata

DongchenZ replied:
Fixed.

mdietze commented on modules/uncertainty/R/generate_joint_ensemble_design.R:
```diff
@@ -95,9 +69,5 @@ generate_joint_ensemble_design <- function(settings,
     sobol_obj <- sensitivity::soboljansen(model = NULL, X1 = X1, X2 = X2)
     return(sobol_obj)
   }
-  # This ensures that regardless of whether the sobol or non-sobol version is called 
```
This comment seems useful, not sure why it was removed

DongchenZ replied:
Reverted.

Commit:
Revert the previous comments and add the parameter generation if we don't have it.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Merge branch 'develop' of https://github.com/PecanProject/pecan into SDA_data

Commit:
Merge branch 'SDA_data' of https://github.com/DongchenZ/pecan into SDA_data

mdietze committed:
Merge branch 'develop' into SDA_data
